Add `goauth` binary for `GOAUTH` environment support

fkollmann opened the pull request:
This adds support for the new `GOAUTH` environment variable, introduced by Go 1.24.

Note: I have not tested the service account support, but it should work. Please feel free to contribute some testing ;) . 

Fixes #21 

# How I tested (binary)

1. Run go build in `cmd/goauth`
2. `export GOAUTH="/home/felix/Projects/artifact-registry-go-tools/cmd/goauth/goauth europe-west1"`
3. `go get -u golang.org/x/net`


# How I tested (script)

1. `export GOAUTH="sh -c 'GOPROXY=direct go run github.com/smartpricer/artifact-registry-go-tools/cmd/goauth@latest europe-west1'"` (obviously using the fork)
2. `go get -u golang.org/x/net`

fkollmann committed:
Add `goauth` binary for `GOAUTH` environment support

owenhaynes commented on cmd/goauth/main.go:
```diff
@@ -0,0 +1,94 @@
+	ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
+	defer cancel()
+
+	host := fmt.Sprintf(hostPattern, location)
```
I may be being a bit thick 

but if 
`GOPROXY=https://europe-west2-go.pkg.dev`. 

Location would be `https://europe-west2-go.pkg.dev` for example this would generate a host of `https://europe-west2-go.pkg.dev-go.pkg.dev` which is invalid.

This should also be testing if the location is ok to be processed by this auth

fkollmann replied:
Hi, sry, took me a while to understand your POV.

The location is the Google Cloud region, as named by the `auth` command.

The second call-case is not needed for the Go proxy authentication here. So this could only happen, if the *location* argument would be missing. 

From my POV the correct way to handle this is to ignore any location which are URLs.

fkollmann replied:
Okay, I understand you are refering to:

> Before the first HTTPS fetch, the go command will invoke each GOAUTH command in the list with no additional arguments and no input. *If the server responds with any 4xx code, the go command will invoke the GOAUTH commands again with the URL as an additional command-line argument and the HTTP Response to the program's stdin.* If the server responds with an error again, the fetch fails: a URL-specific GOAUTH will only be attempted once per fetch.

But the current version is basically ignoring the second case, as it assumes the first one is working. 

I see it is a reasonable extension for security, to prevent sending the authentication to non-Google proxies.

fkollmann replied:
I added a safety check, but from my POV this is nothing that could happen in real life, because this would require for the location argument to be missing, which would then already fail the first call.

owenhaynes replied:
Yeah i misunderstood the location thing as in my version of this arg(0) is the arg go provide on the second invocation
```
If the server responds with any 4xx code, the go command will invoke the
GOAUTH commands again with the URL as an additional command-line argument
and the HTTP Response to the program's stdin.
```

Hence my validation reasoning was behind this.

fkollmann committed:
Change ignoring URLs as location

(for safety for the second call, but this should not happen as this requires the location to be missing, which would fail on the first call.)

yihanzhen commented on cmd/goauth/main.go:
```diff
@@ -0,0 +1,99 @@
+
+  export GOAUTH="sh -c 'GOPROXY=direct go run github.com/GoogleCloudPlatform/artifact-registry-go-tools/cmd/goauth@latest <location>'"
+
+To support multiple locations, add the command multiple times to the GOAUTH variable (semicolon-separated).
```
Can we add an example for using this together with a non Artifact Registry repo? e.g.,

```
To support an Artifact Registry repo together with other repos, add all commands to the GOAUTH variable. For example:

export GOAUTH="GOPROXY=direct go run github.com/GoogleCloudPlatform/artifact-registry-go-tools/cmd/goauth@latest <location>; another command"

fkollmann replied:
I am not against it, but your example won't work, see below. Please feel free to do some testing and verification on your side.

yihanzhen commented on cmd/goauth/main.go:
```diff
@@ -0,0 +1,99 @@
+
+Add to your GOAUTH environment variable:
+
+  export GOAUTH="sh -c 'GOPROXY=direct go run github.com/GoogleCloudPlatform/artifact-registry-go-tools/cmd/goauth@latest <location>'"
```
I wonder if we can leave out `sh -c`. Without it the command should still work but not sure if `sh -c` works on Windows?

cagataygurturk replied:
It won't work on Windows and I doubt sh -c is also needed.

```suggestion
  export GOAUTH="go run github.com/GoogleCloudPlatform/artifact-registry-go-tools/cmd/goauth@latest <location>"
```

must work allright

fkollmann replied:
> It won't work on Windows and I doubt sh -c is also needed.
> 
> must work allright

No, it does not. Please feel free to do some testing and verification.

I already checked these scenarios:

```sh
export GOAUTH="go run github.com/smartpricer/artifact-registry-go-tools/cmd/goauth@latest europe-west1"

#--> recursive call

export GOAUTH="GOPROXY=direct go run github.com/smartpricer/artifact-registry-go-tools/cmd/goauth@latest europe-west1"

#--> 401, does not work

export GOAUTH="sh -c 'GOPROXY=direct go run github.com/smartpricer/artifact-registry-go-tools/cmd/goauth@latest europe-west1'"

#--> works fine
```

fkollmann replied:
Regarding the Windows support: Feel free to suggest an example for Windows, after testing it.

cagataygurturk replied:
Maybe the instructions must change to ask users to run go install first instead of go run

If you install the binary first like this:

`GOPROXY=direct go install github.com/GoogleCloudPlatform/artifact-registry-go-tools/cmd/goauth@latest`

Then you can use the installed binary directly in GOAUTH without having to use GOPROXY

GOAUTH="goauth europe-west1"

owenhaynes replied:
Even better would be for this to be distributed with the gcloud SDK like the docker Auth does.

fkollmann replied:
I think if going with @cagataygurturk , it needs to be renamed, e.g. `go-auth-google-artifact-registry` or `go-auth-gar` or sth. like this.

Besides that, I kind of agree with @owenhaynes , that this should become part of the Google Cloud SDK.

On our side, we solved this easily, because we run a mono repository, so we already have a place to put team-wide tooling.

yihanzhen replied:
> Even better would be for this to be distributed with the gcloud SDK like the docker Auth does.

I had the same thought - but the original rationale of implementing auth helpers in the same language as the repository format is that the users won't have a dependency on gcloud SDK in their build flow, and some customers do care about this mostly because gcloud is huge.

That being said I think implementing it within gcloud is indeed simpler; I'll go ahead and do that first to hopefully unblock folks. I would like to eventually add support for it in this auth tool, but I think it would require some design review among the team so I'll get back to it later.

yihanzhen replied:
FYI after spending a lot of time debugging what's described in https://github.com/golang/go/issues/71889 I finally have made it to work locally. Need some extra time adding tests and stuff and this should be available very soon.

yihanzhen commented on cmd/goauth/main.go:
```diff
@@ -0,0 +1,99 @@
+For more details, see https://pkg.go.dev/cmd/go@master#hdr-GOAUTH_environment_variable`
+
+func main() {
+	jsonKey := flag.String("json_key", "", "path to the json key of the service account used for this location. Leave empty to use the oauth token instead.")
```
I was wondering if we could reuse the same program entry point. The main motivation is to avoid duplicate code (the two flags here), or at least make it harder for future developers to miss one another when adding more features to goauth or netrc.

maybe

```
export GOAUTH="GOPROXY=direct go run github.com/GoogleCloudPlatform/artifact-registry-go-tools/cmd/auth@latest goauth ---location"
```

fkollmann replied:
I think it's better to have it separate, because the application behavior is dictated by Go and will have to be different from the `auth` command behavior: Things like writing and reading from stdin/out and getting an additional commandline argument.

Even if they seem to be similar now, they probably will divert more in the future.

yihanzhen commented on cmd/goauth/main.go:
```diff
@@ -0,0 +1,99 @@
+	}
+}
+
+func handleLocation(location string, jsonKeyPath string, hostPattern string) error {
```
can we rename this to `goauth` or something similar? handleLocation seems a bit vague and away from what the function actually does.

yihanzhen replied:
Can we add a unit test for this function? Maybe also change this function to return `string, error` and outside it write the result to STDOUT, so that it's easier to test.

yihanzhen reviewed (commented):
Really appreciate the contribution. Sorry if I am being a little nit picky here; mostly wanted to share some thoughts from  the maintainability perspective.

Happy to discuss over the comments.

fkollmann committed:
Change help to be written to stderr

fkollmann committed:
Add unit testing

fkollmann commented:
fyi, I submitted a change to Go for adding environment variable support: https://go-review.googlesource.com/c/go/+/650055 

This should make `sh -c` obsolete and improve Windows compatibility.

yihanzhen commented:
Hey folks watching this PR, the gcloud implementation of this request should be out now (still hidden but should be usable). The command is `gcloud artifacts go auth --location=us-west1` or whatever location your Go repo is in. Would anyone be willing to test it out and let me know if it works?

nvx commented:
> Hey folks watching this PR, the gcloud implementation of this request should be out now (still hidden but should be usable). The command is `gcloud artifacts go auth --location=us-west1` or whatever location your Go repo is in. Would anyone be willing to test it out and let me know if it works?

I'm not sure if this is the best place for this, but I wasn't able to spot anything else.

Setting GOAUTH as described it doesn't seem to work, getting

```
go: module golang.org/x/tools/cmd/stringer: reading https://australia-southeast1-go.pkg.dev/<project>/<repo>/golang.org/x/tools/cmd/stringer/@v/list: 401 Unauthorized
        server response: The request does not have valid authentication credentials. If you are running github.com/GoogleCloudPlatform/artifact-registry-go-tools, make sure to add it to the GONOPROXY environment variable. Refer to https://go.dev/ref/mod#private-module-privacy for more information.
```

GOAUTH config is:

```
$ go env GOAUTH
gcloud artifacts go auth --location=australia-southeast1
```

Looking at the docs here https://pkg.go.dev/cmd/go#hdr-GOAUTH_environment_variable it looks like the URL line should end in a slash while the output doesn't seem to be?

```
$ gcloud artifacts go auth --location=australia-southeast1
https://australia-southeast1-go.pkg.dev

Authorization: Bearer <snip>

```

As an aside the first time I ran it I noticed it sasked about the gcloud survey which I imagine would break things, I turned it off locally with `gcloud config set survey/disable_prompts True` which didn't resolve the issue but did stop the survey prompt from showing. This command probably needs to be modified to never ask about the survey but at least there's a workaround.

```
$ gcloud artifacts go auth --location=australia-southeast1
https://australia-southeast1-go.pkg.dev

Authorization: Bearer <snip>



To take a quick anonymous survey, run:
  $ gcloud survey


$
```

antoinedeschenes commented:
> This command probably needs to be modified to never ask about the survey but at least there's a workaround.

`gcloud --quiet artifacts go auth --location=<location>` would make more sense as it disables interactive prompts

```
     --quiet, -q
        Disable all interactive prompts when running gcloud commands. If input
        is required, defaults will be used, or an error will be raised.

        Overrides the default core/disable_prompts property value for this
        command invocation. This is equivalent to setting the environment
        variable CLOUDSDK_CORE_DISABLE_PROMPTS to 1.
```

It's pretty annoying that gcloud won't automatically handle this when it's not running in a tty.

---

Back on topic, I still think it makes sense to support GOAUTH in the standalone artifact-registry-go-tools helper as it doesn't require to download the whole gcloud CLI
